File: 백준/Silver/12852. 1로 만들기 2/1로 만들기 2.cc
#include <iostream>
#include <algorithm>
using namespace std;

int n;
int dp[1000001];

void go(int num) {
	if (num < 1) return;
	cout << num << " ";
	if (num % 3 == 0 && dp[num / 3] + 1 == dp[num]) go(num / 3);
	else if (num % 2 == 0 && dp[num / 2] + 1 == dp[num]) go(num / 2);
	else if (dp[num - 1]+1 == dp[num])go(num - 1);
}

int main() {
	cin >> n;
	fill(dp, dp + 1000001, 987654321);
	dp[1] = 0;
	for (int i = 1; i <= n; i++) {
		if (i % 3 == 0)dp[i] = min(dp[i / 3] + 1, dp[i]);
		if (i % 2 == 0)dp[i] = min(dp[i / 2] + 1, dp[i]);
		dp[i] = min(dp[i - 1]+1, dp[i]);
	}
	cout << dp[n] << '\n';
	go(n);
}

File: 백준/Silver/12852. 1로 만들기 2/1로 만들기 2.java
import java.util.Arrays;
import java.util.Scanner;

public class Main{
    static int n;
    static int[] visited = new int[1000002];

    public static void go(int num){
        if(num < 1) return;
        System.out.print(num + " ");
        if (num % 3 == 0 && visited[num / 3] + 1 == visited[num]) go(num / 3);
        else if (num % 2 == 0 && visited[num / 2] + 1 == visited[num]) go(num / 2);
        else if (visited[num - 1]+1 == visited[num])go(num - 1);
    }

    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        Arrays.fill(visited,10000000);
        visited[1] = 0;
        int temp = 1;
        for (int i = 1; i <= n; i++) {
            if (i % 3 == 0)visited[i] = Math.min(visited[i / 3] + 1, visited[i]);
            if (i % 2 == 0)visited[i] = Math.min(visited[i / 2] + 1, visited[i]);
            visited[i] = Math.min(visited[i - 1]+1, visited[i]);
        }
        System.out.println(visited[n]);
        go(n);
    }
}